lambdas/tabular_preview/src/t4_lambda_tabular_preview/__init__.py

```diff
@@ -215,23 +225,96 @@ def preview_parquet(url, compression, max_out_size):
+                {
+                    "truncated": output_truncated,
+                    "meta": {
+                        "created_by": meta.created_by,
+                        "format_version": meta.format_version,
+                        "num_row_groups": meta.num_row_groups,
+                        "schema": {"names": meta.schema.names},
+                        "serialized_size": meta.serialized_size,
+                        "shape": (meta.num_rows, meta.num_columns),
+                    },
+                }
+            ),
+        },
+    )
+
+
+def preview_h5ad(url, compression, max_out_size):
+    with urlopen(url, compression=compression, seekable=True) as src:
+        with h5py.File(src, "r") as h5py_file:
+            adata = anndata.experimental.read_lazy(h5py_file)
+
+            # Get matrix dimensions to decide processing strategy
+            n_obs, n_vars = adata.shape
+
+            if meta_only := (n_obs * n_vars >= H5AD_META_ONLY_SIZE):
+                # For large files, skip intensive QC calculation that requires loading full matrix
+                logger.warning("Getting only meta for large matrix (%d x %d) to avoid OOM/timeout", n_obs, n_vars)
+
+                # Create empty dataframe
+                var_df = pandas.DataFrame(columns=list(adata.var.keys()))
+            else:
+                adata = anndata.read_h5ad(src)
+                # For smaller matrices, calculate full QC metrics using scanpy
+                sc.pp.calculate_qc_metrics(adata, percent_top=None, log1p=False, inplace=True)
+                var_df = adata.var.copy()
+
+            # Add gene expression statistics from the QC metrics
+            # These columns are added by calculate_qc_metrics:
+            # - total_counts: total UMI counts for this gene across all cells
+            # - n_cells_by_counts: number of cells with non-zero counts for this gene
+            # - mean_counts: mean counts per cell for this gene
+            # - pct_dropout_by_counts: percentage of cells with zero counts
+
+            # Reset index to include gene IDs as a regular column
+            var_df_with_index = var_df.reset_index()
+            # XXX: doesn't that change the original column name?
+            var_df_with_index = var_df_with_index.rename(columns={"index": "gene_id"})
+
+            # Convert to Arrow table and write as Arrow format
+            table = pyarrow.Table.from_pandas(var_df_with_index, preserve_index=False)
+            output_data, output_truncated = write_data_as_arrow(table, table.schema, max_out_size)
+
+    return (
+        200,
+        output_data,
+        {
+            "Content-Type": "application/vnd.apache.arrow.file",
+            "Content-Encoding": "gzip",
+            QUILT_INFO_HEADER: json.dumps(
+                {
+                    "truncated": output_truncated,
+                    "meta_only": meta_only,
+                    # H5AD-specific metadata format
```

Comment:
Perhaps, these comments are redundant (and the one about converting to arrow format)